presto-main/src/main/java/com/facebook/presto/operator/WindowInfo.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.operator.window.WindowPartition;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.concurrent.Immutable;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Iterables.concat;
+
+public class WindowInfo
+        implements Mergeable<WindowInfo>, OperatorInfo
+{
+    private final List<DriverWindowInfo> windowInfos;
+
+    @JsonCreator
+    public WindowInfo(@JsonProperty("windowInfos") List<DriverWindowInfo> windowInfos)
+    {
+        this.windowInfos = ImmutableList.copyOf(windowInfos);
+    }
+
+    @JsonProperty
+    public List<DriverWindowInfo> getWindowInfos()
+    {
+        return windowInfos;
+    }
+
+    @Override
+    public WindowInfo mergeWith(WindowInfo other)
+    {
+        return new WindowInfo(ImmutableList.copyOf(concat(this.windowInfos, other.windowInfos)));
+    }
+
+    @ThreadSafe
+    static class DriverWindowInfoBuilder
+    {
+        private ImmutableList.Builder<IndexInfo> indexInfosBuilder = ImmutableList.builder();
+        private IndexInfoBuilder currentIndexInfoBuilder = null;
+
+        public synchronized void addIndex(PagesIndex index)
+        {
+            if (currentIndexInfoBuilder != null) {
+                Optional<IndexInfo> indexInfo = currentIndexInfoBuilder.build();
+                if (indexInfo.isPresent()) {
+                    indexInfosBuilder.add(indexInfo.get());
+                }
+            }
+            currentIndexInfoBuilder = new IndexInfoBuilder(index.getPositionCount(), index.getEstimatedSize().toBytes());
+        }
+
+        public synchronized void addPartition(WindowPartition partition)
+        {
+            checkState(currentIndexInfoBuilder != null, "addIndex must be called before addPartition");
+            currentIndexInfoBuilder.addPartition(partition);
+        }
+
+        public synchronized DriverWindowInfo build()
+        {
+            if (currentIndexInfoBuilder != null) {
+                Optional<IndexInfo> indexInfo = currentIndexInfoBuilder.build();
```

Comment:
While in `build()` method, we have to nullify `currentIndexInfoBuilder` in addition to adding index info to the builder.
I don't think it brings any more good if I extract this two lines to a separate method:
```java
                Optional<IndexInfo> indexInfo = currentIndexInfoBuilder.build();
                indexInfo.ifPresent(indexInfosBuilder::add);
```
given the `if` cannot be extracted because I need to nullify the `currentIndexInfoBuilder` in `build()` only.